# Exercise 02 -  OLAP Cubes - Grouping Sets

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [1]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

createdb: database creation failed: ERROR:  database "pagila_star" already exists
 set_config 
------------
 
(1 row)

psql:Data/pagila-star.sql:41: ERROR:  type "mpaa_rating" already exists
psql:Data/pagila-star.sql:51: ERROR:  type "year" already exists
psql:Data/pagila-star.sql:68: ERROR:  function "_group_concat" already exists with same argument types
psql:Data/pagila-star.sql:85: ERROR:  function "film_in_stock" already exists with same argument types
psql:Data/pagila-star.sql:102: ERROR:  function "film_not_in_stock" already exists with same argument types
psql:Data/pagila-star.sql:147: ERROR:  function "get_customer_balance" already exists with same argument types
psql:Data/pagila-star.sql:169: ERROR:  function "inventory_held_by_customer" already exists with same argument types
psql:Data/pagila-star.sql:206: ERROR:  function "inventory_in_stock" already exists with same argument types
psql:Data/pagila-star.sql:224: ERROR:  function "last_day" already exists with same argument 

 setval 
--------
  32098
(1 row)

psql:Data/pagila-star.sql:81396: ERROR:  duplicate key value violates unique constraint "rental_pkey"
DETAIL:  Key (rental_id)=(2) already exists.
CONTEXT:  COPY rental, line 1
 setval 
--------
  16049
(1 row)

psql:Data/pagila-star.sql:81413: ERROR:  duplicate key value violates unique constraint "staff_pkey"
DETAIL:  Key (staff_id)=(1) already exists.
CONTEXT:  COPY staff, line 1
 setval 
--------
      2
(1 row)

psql:Data/pagila-star.sql:81430: ERROR:  duplicate key value violates unique constraint "store_pkey"
DETAIL:  Key (store_id)=(1) already exists.
CONTEXT:  COPY store, line 1
 setval 
--------
      2
(1 row)

psql:Data/pagila-star.sql:81445: ERROR:  multiple primary keys for table "actor" are not allowed
psql:Data/pagila-star.sql:81453: ERROR:  multiple primary keys for table "address" are not allowed
psql:Data/pagila-star.sql:81461: ERROR:  multiple primary keys for table "category" are not allowed
psql:Data/pagila-star.sql:81469: ERROR:

### Connect to the local database where Pagila is loaded

In [2]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

# Grouping Sets
- It happens often that for 3 dimensions, you want to aggregate a fact:
    - by nothing (total)
    - then by the 1st dimension
    - then by the 2nd 
    - then by the 3rd 
    - then by the 1st and 2nd
    - then by the 2nd and 3rd
    - then by the 1st and 3rd
    - then by the 1st and 2nd and 3rd
    
- Since this is very common, and in all cases, we are iterating through all the fact table anyhow, there is a more clever way to do that using the SQL grouping statement "GROUPING SETS" 

## Total Revenue

TODO: Write a query that calculates total revenue (sales_amount)

In [5]:
%%sql
SELECT sum(sales_amount) as revenue
FROM factSales 

 * postgresql://student:***@127.0.0.1:5432/pagila_star
1 rows affected.


[(Decimal('67416.51'),)]

## Revenue by Country
TODO: Write a query that calculates total revenue (sales_amount) by country

In [9]:
%%sql
SELECT ct.country, sum(fs.sales_amount) as revenue
FROM factSales  fs
JOIN dimCustomer ct ON ct.customer_key = fs.customer_key
GROUP BY ct.country
ORDER BY ct.country, revenue desc


 * postgresql://student:***@127.0.0.1:5432/pagila_star
108 rows affected.


[('Afghanistan', Decimal('67.82')),
 ('Algeria', Decimal('383.10')),
 ('American Samoa', Decimal('71.80')),
 ('Angola', Decimal('215.48')),
 ('Anguilla', Decimal('106.65')),
 ('Argentina', Decimal('1434.48')),
 ('Armenia', Decimal('118.75')),
 ('Austria', Decimal('307.22')),
 ('Azerbaijan', Decimal('245.43')),
 ('Bahrain', Decimal('112.75')),
 ('Bangladesh', Decimal('402.05')),
 ('Belarus', Decimal('277.34')),
 ('Bolivia', Decimal('183.53')),
 ('Brazil', Decimal('3200.52')),
 ('Brunei', Decimal('113.65')),
 ('Bulgaria', Decimal('204.50')),
 ('Cambodia', Decimal('191.47')),
 ('Cameroon', Decimal('200.46')),
 ('Canada', Decimal('593.63')),
 ('Chad', Decimal('135.68')),
 ('Chile', Decimal('328.29')),
 ('China', Decimal('5802.73')),
 ('Colombia', Decimal('709.41')),
 ('Congo, The Democratic Republic of the', Decimal('212.50')),
 ('Czech Republic', Decimal('133.71')),
 ('Dominican Republic', Decimal('318.23')),
 ('Ecuador', Decimal('390.13')),
 ('Egypt', Decimal('694.39')),
 ('Estonia', Decimal('115.70')),
 ('Ethiopia', Decimal('91.77')),
 ('Faroe Islands', Decimal('114.72')),
 ('Finland', Decimal('101.74')),
 ('France', Decimal('374.04')),
 ('French Guiana', Decimal('103.78')),
 ('French Polynesia', Decimal('235.46')),
 ('Gambia', Decimal('129.70')),
 ('Germany', Decimal('831.04')),
 ('Greece', Decimal('232.46')),
 ('Greenland', Decimal('137.66')),
 ('Holy See (Vatican City State)', Decimal('152.66')),
 ('Hong Kong', Decimal('142.70')),
 ('Hungary', Decimal('111.71')),
 ('India', Decimal('6630.27')),
 ('Indonesia', Decimal('1510.33')),
 ('Iran', Decimal('950.75')),
 ('Iraq', Decimal('111.73')),
 ('Israel', Decimal('422.01')),
 ('Italy', Decimal('831.11')),
 ('Japan', Decimal('3471.74')),
 ('Kazakstan', Decimal('198.48')),
 ('Kenya', Decimal('253.46')),
 ('Kuwait', Decimal('111.74')),
 ('Latvia', Decimal('262.40')),
 ('Liechtenstein', Decimal('114.72')),
 ('Lithuania', Decimal('73.76')),
 ('Madagascar', Decimal('93.78')),
 ('Malawi', Decimal('123.72')),
 ('Malaysia', Decimal('369.15')),
 ('Mexico', Decimal('3307.04')),
 ('Moldova', Decimal('127.66')),
 ('Morocco', Decimal('307.29')),
 ('Mozambique', Decimal('344.20')),
 ('Myanmar', Decimal('194.48')),
 ('Nauru', Decimal('148.69')),
 ('Nepal', Decimal('116.78')),
 ('Netherlands', Decimal('586.66')),
 ('New Zealand', Decimal('92.76')),
 ('Nigeria', Decimal('1511.48')),
 ('North Korea', Decimal('113.69')),
 ('Oman', Decimal('187.50')),
 ('Pakistan', Decimal('525.72')),
 ('Paraguay', Decimal('275.38')),
 ('Peru', Decimal('468.88')),
 ('Philippines', Decimal('2381.32')),
 ('Poland', Decimal('877.97')),
 ('Puerto Rico', Decimal('254.39')),
 ('Romania', Decimal('235.38')),
 ('Runion', Decimal('216.54')),
 ('Russian Federation', Decimal('3045.87')),
 ('Saint Vincent and the Grenadines', Decimal('96.75')),
 ('Saudi Arabia', Decimal('523.79')),
 ('Senegal', Decimal('100.75')),
 ('Slovakia', Decimal('89.74')),
 ('South Africa', Decimal('1204.15')),
 ('South Korea', Decimal('574.65')),
 ('Spain', Decimal('606.58')),
 ('Sri Lanka', Decimal('116.70')),
 ('Sudan', Decimal('227.46')),
 ('Sweden', Decimal('144.66')),
 ('Switzerland', Decimal('252.39')),
 ('Taiwan', Decimal('1210.94')),
 ('Tanzania', Decimal('341.17')),
 ('Thailand', Decimal('419.04')),
 ('Tonga', Decimal('73.82')),
 ('Tunisia', Decimal('78.77')),
 ('Turkey', Decimal('1662.12')),
 ('Turkmenistan', Decimal('136.73')),
 ('Tuvalu', Decimal('120.74')),
 ('Ukraine', Decimal('730.42')),
 ('United Arab Emirates', Decimal('333.16')),
 ('United Kingdom', Decimal('924.80')),
 ('United States', Decimal('4110.32')),
 ('Venezuela', Decimal('683.30')),
 ('Vietnam', Decimal('746.28')),
 ('Virgin Islands, U.S.', Decimal('122.68')),
 ('Yemen', Decimal('510.83')),
 ('Yugoslavia', Decimal('259.43')),
 ('Zambia', Decimal('134.67'))]

## Revenue by Month
TODO: Write a query that calculates total revenue (sales_amount) by month

In [10]:
%%sql
SELECT dt.month, sum(fs.sales_amount) as revenue
FROM factSales  fs
JOIN dimDate dt ON dt.date_key = fs.date_key
GROUP BY dt.month
ORDER BY dt.month, revenue desc

 * postgresql://student:***@127.0.0.1:5432/pagila_star
5 rows affected.


[(1, Decimal('4824.43')),
 (2, Decimal('9631.88')),
 (3, Decimal('23886.56')),
 (4, Decimal('28559.46')),
 (5, Decimal('514.18'))]

## Revenue by Month & Country
TODO: Write a query that calculates total revenue (sales_amount) by month and country. Sort the data by month, country, and revenue in descending order. The first few rows of your output should match the table below.

In [16]:
%%sql
SELECT dt.month, ct.country, sum(fs.sales_amount) as revenue
FROM factSales  fs
JOIN dimDate dt ON dt.date_key = fs.date_key
JOIN dimStore ct ON ct.store_key = fs.store_key
GROUP BY (dt.month, ct.country)
ORDER BY dt.month, ct.country, revenue desc;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
10 rows affected.


[(1, 'Australia', Decimal('2364.19')),
 (1, 'Canada', Decimal('2460.24')),
 (2, 'Australia', Decimal('4895.10')),
 (2, 'Canada', Decimal('4736.78')),
 (3, 'Australia', Decimal('12060.33')),
 (3, 'Canada', Decimal('11826.23')),
 (4, 'Australia', Decimal('14136.07')),
 (4, 'Canada', Decimal('14423.39')),
 (5, 'Australia', Decimal('271.08')),
 (5, 'Canada', Decimal('243.10'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
</tbody></table></div>

## Revenue Total, by Month, by Country, by Month & Country All in one shot

TODO: Write a query that calculates total revenue at the various grouping levels done above (total, by month, by country, by month & country) all at once using the grouping sets function. Your output should match the table below.

In [17]:
%%sql
SELECT dt.month, ct.country, sum(fs.sales_amount) as revenue
FROM factSales  fs
JOIN dimDate dt ON dt.date_key = fs.date_key
JOIN dimStore ct ON ct.store_key = fs.store_key
GROUP BY grouping sets((), dt.month, ct.country, (dt.month, ct.country))


 * postgresql://student:***@127.0.0.1:5432/pagila_star
18 rows affected.


[(1, 'Australia', Decimal('2364.19')),
 (1, 'Canada', Decimal('2460.24')),
 (1, None, Decimal('4824.43')),
 (2, 'Australia', Decimal('4895.10')),
 (2, 'Canada', Decimal('4736.78')),
 (2, None, Decimal('9631.88')),
 (3, 'Australia', Decimal('12060.33')),
 (3, 'Canada', Decimal('11826.23')),
 (3, None, Decimal('23886.56')),
 (4, 'Australia', Decimal('14136.07')),
 (4, 'Canada', Decimal('14423.39')),
 (4, None, Decimal('28559.46')),
 (5, 'Australia', Decimal('271.08')),
 (5, 'Canada', Decimal('243.10')),
 (5, None, Decimal('514.18')),
 (None, None, Decimal('67416.51')),
 (None, 'Australia', Decimal('33726.77')),
 (None, 'Canada', Decimal('33689.74'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>1</td>
        <td>None</td>
        <td>4824.43</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>2</td>
        <td>None</td>
        <td>9631.88</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Canada</td>
        <td>11826.23</td>
    </tr>
    <tr>
        <td>3</td>
        <td>None</td>
        <td>23886.56</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Australia</td>
        <td>14136.07</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Canada</td>
        <td>14423.39</td>
    </tr>
    <tr>
        <td>4</td>
        <td>None</td>
        <td>28559.46</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Australia</td>
        <td>271.08</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Canada</td>
        <td>243.10</td>
    </tr>
    <tr>
        <td>5</td>
        <td>None</td>
        <td>514.18</td>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>67416.51</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Australia</td>
        <td>33726.77</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Canada</td>
        <td>33689.74</td>
    </tr>
</tbody></table></div>